include/fiction/algorithms/physical_design/on_the_fly_circuit_design_on_defective_surface.hpp

```diff
@@ -0,0 +1,216 @@
+                        *gate_level_layout, params.parameterized_gate_library_parameter);
+                }
+
+                // on-the-fly gate design was unsuccessful at a certain tile. Hence, this tile-gate pair is added to the
+                // blacklist.
+                catch (const gate_design_exception<tt, GateLyt>& e)
+                {
+                    black_list[e.which_tile()][e.which_truth_table()].push_back(e.which_port_list());
+                }
+
+                catch (const std::exception& e)
+                {
+                    std::cerr << "Caught std::exception: " << e.what() << '\n';
+                }
+            }
+
+            // P&R was unsuccessful
+            else
+            {
+                break;
+            }
+        }
+
+        // empty defect-surface is returned when no P&R solution is found
+        if (!gate_level_layout.has_value())
+        {
+            return sidb_defect_surface<CellLyt>{};
+        }
+
+        // check equivalence
+        const auto miter = mockturtle::miter<technology_network>(mapped_network, *gate_level_layout);
+        const auto eq    = mockturtle::equivalence_checking(*miter);
+
+        // empty defect-surface is returned when equivalence check is empty
+        if (!eq.has_value())
+        {
+            return sidb_defect_surface<CellLyt>{};
+        }
+
+        // in case of equality, an sidb_defect_surface with the SiDBs of the circuit is returned
+        if (*eq)
+        {
+            sidb_defect_surface<CellLyt> sidbs_and_defects{cell_level_layout};
+
+            // add defects to the circuit.
+            params.parameterized_gate_library_parameter.defect_surface.foreach_sidb_defect(
+                [&sidbs_and_defects](const auto& defect)
+                { sidbs_and_defects.assign_sidb_defect(defect.first, defect.second); });
+
+            return sidbs_and_defects;
+        }
+
+        // in case of inequality, empty defect-surface is returned
+        return sidb_defect_surface<CellLyt>{};
+    }
+
+  private:
+    /**
+     * Gate-level layout.
+     */
+    GateLyt lattice_tiling;
+    /**
+     * Mapped network.
+     */
+    Ntk mapped_network{};
+    /**
+     * Parameters for the on-the-fly circuit design.
+     */
+    on_the_fly_circuit_design_params<CellLyt> params{};
+    /**
+     * Statistics for the on-the-fly circuit design.
+     */
+    on_the_fly_circuit_design_stats& stats;
+};
+
+}  // namespace detail
+
+/**
+ * This function applies the on-the-fly circuit design algorithm to a specified defective surface, automatically
+ * creating customized SiDB gates while integrating atomic defects as essential components of the design. It was
+ * proposed in \"On-the-fly Defect-Aware Design of Circuits based on Silicon Dangling Bond Logic\" by J. Drewniok, M.
+ * Walter, S. S. H. Ng, K. Walus, and R. Wille in IEEE NANO 2024.
```

Comment:
How does this work exactly? What happens inside the algorithm? See, e.g., `exact.hpp` or `operational_domain.hpp`